lib/upload.js

```diff
@@ -923,15 +924,24 @@ function shouldRetry(err, retryAttempt, options) {
   // - retryDelays option is set
   // - we didn't exceed the maxium number of retries, yet, and
   // - this error was caused by a request or it's response and
-  // - the error is server error (i.e. no a status 4xx or a 409 or 423) and
+  // - the error is server error (i.e. no a status 4xx or a 409 or 423) or
+  // a onShouldRetry is specified and returns true
   // - the browser does not indicate that we are offline
   let status = err.originalResponse ? err.originalResponse.getStatus() : 0;
-  let isServerError = !inStatusCategory(status, 400) || status === 409 || status === 423;
+
+  // define as function so we only call it if really needed
+  let doRetry = () => {
+    if (options && typeof options.onShouldRetry === "function") {
+      return options.onShouldRetry(err, retryAttempt, options);
+    } else {
+      return !inStatusCategory(status, 400) || status === 409 || status === 423;
+    }
+  };
   return options.retryDelays != null &&
          retryAttempt < options.retryDelays.length &&
          err.originalRequest != null &&
-         isServerError &&
-         isOnline();
+         isOnline() &&
```

Comment:
If we already expose a generic function for determining retries, I think we should also remove the online check if the user supplied onShouldRetry. There are use cases where people want to overwrite that check, I believe. What do you think?